nixos/modules/misc/nixpkgs.nix

```diff
@@ -280,6 +280,11 @@ in
     system = mkOption {
       type = types.str;
       example = "i686-linux";
+      default =
+        throwIf
+          (opt.system.highestPrio == (lib.mkOptionDefault {}).priority)
+          ''The option ${opt.system} has not been set. Please set ${opt.hostPlatform} instead. Note that nixos-generate-config will set it in hardware-configuration.nix.''
```

Comment:
Setting `system` is still valid for now, as we can't start an ecosystem-wide change without first making the new way widely available. Even adding a warning will cause annoyance as 2nd-party repos can not implement the change without breaking 22.05 support. So this will take time. We can consider this for 23.05, and unstable in January when 22.05 is EOL.

I've put in place migration logic for such changes, but I didn't use it here because the implementation is complicated enough as it is. See https://github.com/NixOS/release-wiki/issues/31.
We could add such a warning in advance with `isInOldestRelease`, but I'm hesitant to automate this before this has ecosystem support.

Reply:
I'm mainly asking because of the sentence itself, which doesn't make sense to me. If it's complaining about `system` _not_ being set, then the natural solution would be to _set_ `system`. The sentence only makes sense to me if you remove the first "not".

Reply:
@infinisil Could you review again?

I've taken into account your feedback that it's ok to set `system`.
Setting `hostPlatform` is the default solution for an unset `system` in the case of `nixos-generate-config`-based configurations, so this should be mentioned. Eventually we can deprecate `system` and get rid of this complexity.

The new message is:

`The option nixpkgs.system has not been set. Please set it, or set nixpkgs.hostPlatform instead. Note that recent versions of nixos-generate-config will set the latter for you in hardware-configuration.nix.`

Reply:
Maybe we could explain when somebody would want to use one or the other?  In particular, push people towards setting `hostPlatform` if they don't need to support 22.05?  Otherwise, people will see this message, set `nixpkgs.system` because it's the first thing mentioned, then have to change it _again_ when we deprecate `nixpkgs.system`.  We could save them that trouble by encouraging them to use `hostPlatform` if they don't need 22.05 support.

Reply:
Yeah I was trying to be too conservative (and too concise for that matter). The new message is:

```
error: Neither nixpkgs.hostPlatform nor or the legacy option nixpkgs.system has been set.
       You can set nixpkgs.hostPlatform in hardware-configuration.nix by re-running
       a recent version of nixos-generate-config.
       The option nixpkgs.system is still fully supported for NixOS 22.05 interoperability,
       but will be deprecated in the future, so we recommend to set nixpkgs.hostPlatform.
```